### PR TITLE
[Quantization] Add BitsAndBytes W4 online quantization for diffusion transformer

### PR DESCRIPTION
**Pull request by dpeng123**
RFC #1854
#5091 
## Purpose

Add online BitsAndBytes W4 (NF4, weight-only) quantization for diffusion DiTs, mirroring the existing INT8 online pattern. Enables `--quantization bitsandbytes` for Z-Image and other diffusion models without a pre-quantized checkpoint.

Also fixes a wiring gap in `AsyncOmniEngine`: top-level `quantization_config` was not forwarded to diffusion stage `engine_args`, so callers using the dict API (benchmarks, compare tools, `--ignored-layers`) silently ran BF16. The fix injects `quantization_config` in `_create_default_diffusion_stage_cfg` and `_resolve_stage_configs`, alongside the existing `quantization` string path.

## Test Plan

```bash
# GPU offline end-to-end
python benchmarks/diffusion/quantization_quality.py \
  --model Tongyi-MAI/Z-Image-Turbo \
  --task t2i \
  --quantization bitsandbytes \
  --prompts "a cup of coffee on the table" \
  --height 1024 --width 1024 \
  --num-inference-steps 50 \
  --seed 42
```

```bash
# GPU online serving smoke test
vllm serve Tongyi-MAI/Z-Image-Turbo \
  --omni \
  --quantization bitsandbytes \
  --host 127.0.0.1 \
  --port 8092
```

```bash
curl -sS -X POST http://127.0.0.1:8092/v1/images/generations \
  -H "Content-Type: application/json" \
  -d '{
    "prompt": "a cup of coffee on the table",
    "size": "1024x1024",
    "seed": 42
  }'
```

### Offline benchmark

| Item | Value |
| --- | --- |
| GPU | NVIDIA RTX PRO 4500 32GB (Blackwell sm_120) |
| vLLM | 0.24.0 |
| bitsandbytes | 0.49.2 |
| TP | 1 |
| Model | `Tongyi-MAI/Z-Image-Turbo` |
| Setup | 1024x1024, 50 steps, seed 42, LPIPS (alex) |

### Online serving smoke test

| Item | Value |
| --- | --- |
| GPU | NVIDIA RTX PRO 6000 Blackwell Workstation Edition 96GB |
| vLLM image | `vllm-omni_v0.24.0-cuda-12.9` |
| vLLM | 0.24.0 |
| bitsandbytes | 0.49.2 |
| TP | 1 |
| Model | `Tongyi-MAI/Z-Image-Turbo` |
| Endpoint | `POST /v1/images/generations` |
| Request | 1024x1024, seed 42 |

### Offline benchmark

| Config | Avg Time | Speedup | Peak Memory (GiB) | Mem Reduction | Mean LPIPS |
| --- | ---: | ---: | ---: | ---: | ---: |
| BF16 baseline | 45.51s | 1.00x | 23.65 | - | ref |
| bitsandbytes W4 | 46.82s | 0.97x | 9.81 | 59% | 0.0522 |

Visual comparison (prompt: `a cup of coffee on the table`, seed 42):

| BF16 baseline | bitsandbytes W4 |
| --- | --- |
| BF16 baseline | bitsandbytes W4 |

Takeaways:

- Memory: ~59% peak VRAM reduction (23.65 -> 9.81 GiB).
- Speed: No speedup expected for weight-only W4; latency matches BF16 (~46s).
- Quality: LPIPS 0.052 - minor perceptual drift, below the 0.1 "clearly noticeable" threshold. Both outputs are coherent for the test prompt.

### Online serving smoke test

Verified online serving with `--quantization bitsandbytes` using the OpenAI-compatible image generation endpoint.

Server command:

```bash
vllm serve Tongyi-MAI/Z-Image-Turbo \
  --omni \
  --quantization bitsandbytes \
  --host 127.0.0.1 \
  --port 8092
```

Server log evidence:

```text
non-default args: {'model_tag': 'Tongyi-MAI/Z-Image-Turbo', 'host': '127.0.0.1', 'port': 8092, 'model': 'Tongyi-MAI/Z-Image-Turbo', 'quantization': 'bitsandbytes'}
Building quantization config: bitsandbytes
Route: /v1/images/generations, Methods: POST
Application startup complete.
127.0.0.1:44714 - "POST /v1/images/generations HTTP/1.1" 200 OK
```

Request:

```json
{
  "prompt": "a cup of coffee on the table",
  "size": "1024x1024",
  "seed": 42
}
```

Result:

```text
curl_returncode=0
elapsed_sec=17.508
output=/root/pr5037/online_results/bnb_online.png
image_size_bytes=3147861
```

Generated image:

bnb online serving result

Takeaway:

- Online serving works with `vllm serve ... --quantization bitsandbytes`.
- The `/v1/images/generations` endpoint successfully returned a generated image.
- The server log confirms the BitsAndBytes quantization config was built on the online serving path.

BEFORE SUBMITTING: read `CONTRIBUTING.md` and run the `precheck-pr` skill with the code agent for a self-check against project conventions.
(anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

**Comment by lishunyang12** [comment]
Have you tried online serving cases?

**Comment by dpeng123** [comment]
> Have you tried online serving cases?

@lishunyang12 Tested online serving cases, it works, put in PR description, could you help check? Thanks!

**Comment by dpeng123** [comment]
@lishunyang12 gently reminder for review, thanks!

**Comment by hsliuustc0106** [comment]
@david6666666 @Isotr0py PTAL

File: tests/diffusion/quantization/test_bitsandbytes_config.py [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+        config.quant_type = "nf4"
+        config.compress_statistics = True
+        return config
+
+    @pytest.fixture
+    def mock_deps(self, mocker):
+        mock_qweight = torch.ones((64, 32), dtype=torch.uint8)
+        mock_quant_state = mocker.Mock()
+        mock_quant = mocker.patch(
+            "bitsandbytes.functional.quantize_4bit",
```
Comment (david6666666) [inline tests/diffusion/quantization/test_bitsandbytes_config.py:141]: This fixture imports bitsandbytes through mocker.patch, but TestBnBOnlineLinearMethod is not guarded by bitsandbytes_available. Core-model tests will fail with ModuleNotFoundError on installations without this optional package; please guard this class/fixture or inject a fake module.

Comment (dpeng123) [reply tests/diffusion/quantization/test_bitsandbytes_config.py:141]: Thanks for reviewing the PR, updated PR to address the comment, could you help check?

Comment (david6666666) [reply tests/diffusion/quantization/test_bitsandbytes_config.py:141]: The optional-dependency issue is fixed now, thanks.

File: tests/diffusion/quantization/test_bitsandbytes_config.py [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+            "mock_quant_state": mock_quant_state,
+        }
+
+    def test_process_weights_after_loading(self, mock_deps, mock_quant_config, mocker):
+        from vllm_omni.quantization.bitsandbytes_config import BnBOnlineLinearMethod
+
+        method = BnBOnlineLinearMethod(mock_quant_config)
+        layer = Module()
+        layer.weight = Parameter(torch.randn(64, 32))
+        mocker.patch.object(torch.Tensor, "cuda", side_effect=lambda self, *args, **kwargs: self)
```
Comment (david6666666) [inline tests/diffusion/quantization/test_bitsandbytes_config.py]: This still fails on the latest head: patching Tensor.cuda this way does not bind self, so process_weights_after_loading raises TypeError. I get 10 passed, 1 failed with --run-level core_model. Please use autospec=True or return the expected tensor directly.

File: vllm_omni/engine/async_omni_engine.py [hunk 3]
```diff
@@ -1241,6 +1241,13 @@ def _resolve_stage_configs(self, model: str, kwargs: dict[str, Any]) -> tuple[st
                         or cfg.engine_args.quantization_config is None
                     ):
                         cfg.engine_args.quantization_config = quantization_config
+                top_level_quantization_config = kwargs.get("quantization_config")
```
Comment (david6666666) [inline vllm_omni/engine/async_omni_engine.py]: This block is redundant: the branch immediately above already resolves diffusion_quantization_config or quantization_config and populates the same field. The new regression test also passes against main without these lines. Please remove this no-op branch.

**Comment by david6666666** [comment]
Local verification on the current runtime code with 1x B300, vLLM 0.25.0, bitsandbytes 0.49.2, Z-Image-Turbo, 1024x1024, 50 steps, and seed 42:

- BF16: 5.99 s, 20.08 GiB after model load
- BitsAndBytes: 6.28 s, 7.30 GiB after model load
- LPIPS: 0.3067, versus 0.0522 reported here; the generated samples also have visibly different composition
- Online /v1/images/generations smoke passed with HTTP 200

The memory reduction and serving path are confirmed, but the accuracy result is not reproducible after the vLLM 0.25.0 rebase. Please rerun the A/B on the current head with warmup excluded, at least 3 measured runs and multiple prompts, and document any required ignored_layers. The test-plan command also needs updating: --run-level L1 is no longer valid, and quantization_quality.py currently ends with ZeroDivisionError because the parent process observes zero peak memory.

File: vllm_omni/engine/async_omni_engine.py (hunk 3 above)
Comment (dpeng123) [reply vllm_omni/engine/async_omni_engine.py]: Just notice there was change made to pass quantization_config after I created the PR https://github.com/vllm-project/vllm-omni/pull/4769, it was not there before, removed

File: tests/diffusion/quantization/test_bitsandbytes_config.py (hunk 2 above)
Comment (dpeng123) [reply tests/diffusion/quantization/test_bitsandbytes_config.py]: Fixed by return expected tensor directly.

**Comment by dpeng123** [comment]
> Local verification on the current runtime code with 1x B300, vLLM 0.25.0, bitsandbytes 0.49.2, Z-Image-Turbo, 1024x1024, 50 steps, and seed 42:
> 
> * BF16: 5.99 s, 20.08 GiB after model load
> * BitsAndBytes: 6.28 s, 7.30 GiB after model load
> * LPIPS: 0.3067, versus 0.0522 reported here; the generated samples also have visibly different composition
> * Online /v1/images/generations smoke passed with HTTP 200
> 
> The memory reduction and serving path are confirmed, but the accuracy result is not reproducible after the vLLM 0.25.0 rebase. Please rerun the A/B on the current head with warmup excluded, at least 3 measured runs and multiple prompts, and document any required ignored_layers. The test-plan command also needs updating: --run-level L1 is no longer valid, and quantization_quality.py currently ends with ZeroDivisionError because the parent process observes zero peak memory.

Thanks for the deatiled review, just tested again, here is the result, could you check?
Reran on the current PR head with vLLM 0.25.0 and bitsandbytes 0.49.2: one excluded warmup, 3 measured runs per prompt, 4 prompts, 1024x1024, 50 steps, and seed 42.

| Config | Load memory | Mean generation time | Mean LPIPS |
|---|---:|---:|---:|
| BF16 | 20.83 GiB | 45.96 s | ref |
| BitsAndBytes NF4 | 8.02 GiB | 47.28 s | 0.1905 |

## Per-Prompt LPIPS

Prompt | Mean LPIPS |
|---|---:|
| A high-detail studio photo of an orange tabby cat sitting on a laptop keyboard. | 0.0548 |
| An aerial view of a coral reef with crystal clear turquoise water. | 0.2143 |
| A cinematic night market street scene in Tokyo after rain, neon reflections, 50mm lens. | 0.3549 |
| A macro photograph of a gourmet chocolate mousse dessert with gold leaf on a black plate. | 0.1380

File: vllm_omni/quantization/factory.py [hunk 4]
```diff
@@ -134,6 +141,7 @@ def _build_inc(**kw: Any) -> QuantizationConfig:
 
 _OVERRIDES: dict[str, Callable[..., QuantizationConfig]] = {
     "int8": _build_int8,
+    "bitsandbytes": _build_bitsandbytes,
```
Comment (david6666666) [inline vllm_omni/quantization/factory.py:144]: The PR advertises `bnb` as an alias, but only `bitsandbytes` is registered here: `build_quant_config("bnb")` raises `ValueError`, and the CLI rejects it. Please register the alias consistently or remove it from the PR description.

File: docs/user_guide/quantization/bitsandbytes.md [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+This is an online (dynamic) path: load the normal HuggingFace checkpoint and
+quantize during model loading via the `bitsandbytes` CUDA kernels.
+
+## Hardware Support
+
+| Device | Support |
+|--------|---------|
+| NVIDIA CUDA GPU (SM 75+) | ✅ |
+| NVIDIA Blackwell GPU (SM 100+) | ✅ |
+| NVIDIA Ada/Hopper GPU (SM 89+) | ✅ |
```
Comment (david6666666) [inline docs/user_guide/quantization/bitsandbytes.md:18]: `get_quant_method()` explicitly raises on every non-CUDA platform, so ROCm is unsupported here rather than unverified. Please mark it ❌; also change the example’s “any CUDA GPU” wording to match `get_min_capability() == 75`.

File: vllm_omni/quantization/factory.py (hunk 4 above)
Comment (dpeng123) [reply vllm_omni/quantization/factory.py:144]: removed bnb from pr description

File: docs/user_guide/quantization/bitsandbytes.md (hunk 5 above)
Comment (dpeng123) [reply docs/user_guide/quantization/bitsandbytes.md:18]: fixed

**Comment by david6666666** [comment]
LGTM now. Thank you for your contribution.

**Review by david6666666 — APPROVED** [review]
